File: frontend/src/lib/types/catalog.ts
// Types for the Catalogs surface. Mirrors the JSON returned by the
// `/api/catalogs` endpoints (backend `CatalogResponse` / `CatalogItemResponse`
// / `GuidedBuyingSuggestion`). Money fields arrive as numbers (backend
// `float(...)`); date/datetime fields are ISO strings.
//
// Response money is therefore `MoneyAmount` — honest about the JSON-number
// wire shape while still making `a - b` / `.toFixed()` a type error. Money the
// user TYPES is `MoneyString`: `schemas/catalog.py::CatalogItemCreate` declares
// `unit_price` a `Decimal`, and `json.loads` turns a fractional JSON number
// into a float before any validator can intervene.

import type { BadgeTone } from '$lib/components/ui/Badge.svelte';
import type { MessageKey } from '$lib/i18n/messages';
import type { MoneyAmount, MoneyString } from '$lib/utils/money';

export type CatalogType = 'internal' | 'punchout';

export const CATALOG_TYPES: CatalogType[] = ['internal', 'punchout'];

// The catalog type is a table cell on the (extracted) `/catalogs` list and a
// `<select>` option in `CatalogModal`, both beside translated copy — so it is
// keyed like every other value map in this tree.
export const CATALOG_TYPE_LABEL_KEYS: Record<CatalogType, MessageKey> = {
	internal: 'catalogs.type.internal',
	punchout: 'catalogs.type.punchout'
};

/** The message key for a catalog type, or `null` for an unrecognised one. */
export function catalogTypeLabelKey(catalogType: string): MessageKey | null {
	return CATALOG_TYPE_LABEL_KEYS[catalogType as CatalogType] ?? null;
}

export interface CatalogItem {
	id: string;
	catalog_id: string;
	sku: string | null;
	name: string;
	description: string | null;
	unit_price: MoneyAmount;
	currency: string;
	uom: string | null;
	vendor_id: string | null;
	gl_account_id: string | null;
	category: string | null;
	is_active: boolean;
	created_at: string;
	updated_at: string;
}

export interface Catalog {
	id: string;
	name: string;
	catalog_type: string;
	vendor_id: string | null;
	/** The supplier's NAME, resolved server-side — the same field `Contract` /
	 *  `RecurringTemplate` / `Reconciliation` already carry. `ui/VendorPicker`
	 *  needs it as `selectedLabel`: the chosen vendor can sit on any page of the
	 *  tenant's set, so an id alone leaves the edit form unable to name it. */
	vendor_name: string | null;
	punchout_url: string | null;
	is_active: boolean;
	is_preferred: boolean;
	description: string | null;
	item_count: number;
	items: CatalogItem[];
	created_at: string;
	updated_at: string;
}

export interface CatalogListResponse {
	items: Catalog[];
	total: number;
	page: number;
	page_size: number;
}

// Payload shapes for create / update (request side).
export interface CatalogCreate {
	name: string;
	catalog_type: CatalogType;
	vendor_id: string | null;
	punchout_url: string | null;
	is_active: boolean;
	is_preferred: boolean;
	description: string | null;
}

export interface CatalogItemCreate {
	sku: string | null;
	name: string;
	description: string | null;
	/** Request side — the exact decimal text typed, never a JSON number. */
	unit_price: MoneyString | null;
	currency: string;
	uom: string | null;
	vendor_id: string | null;
	gl_account_id: string | null;
	category: string | null;
	is_active: boolean;
}

// ===================== Guided buying =====================

export interface GuidedBuyingVendor {
	vendor_id: string;
	vendor_name: string;
	reasons: string[]; // 'preferred_catalog' | 'active_contract'
	contract_id: string | null;
	contract_number: string | null;
	catalog_id: string | null;
	catalog_name: string | null;
}

export interface GuidedBuyingItem {
	catalog_item_id: string;
	catalog_id: string;
	catalog_name: string;
	sku: string | null;
	name: string;
	unit_price: MoneyAmount;
	currency: string;
	uom: string | null;
	vendor_id: string | null;
	category: string | null;
	is_preferred: boolean;
}

export interface GuidedBuyingSuggestion {
	preferred_vendors: GuidedBuyingVendor[];
	in_contract_vendors: GuidedBuyingVendor[];
	items: GuidedBuyingItem[];
}

// Why guided buying is steering the buyer at this vendor. The backend's
// vocabulary is fixed but open-ended (`api/catalogs.py` builds the list), so
// the accessor is tolerant: an unrecognised reason renders its raw code rather
// than disappearing from the card — a reason nobody can read is still better
// evidence than no reason at all.
export const GUIDED_BUYING_REASON_LABEL_KEYS: Record<string, MessageKey> = {
	preferred_catalog: 'catalogs.guided.reason.preferredCatalog',
	active_contract: 'catalogs.guided.reason.activeContract'
};

/** The message key for a guided-buying reason, or `null` for an unknown one. */
export function guidedBuyingReasonLabelKey(reason: string): MessageKey | null {
	return GUIDED_BUYING_REASON_LABEL_KEYS[reason] ?? null;
}

// ===================== Punch-out (live cXML/OCI round-trip) =====================

export type PunchoutSessionStatus =
	| 'pending'
	| 'returned'
	| 'converted'
	| 'expired'
	| 'cancelled';

/**
 * The i18n key carrying each session-status label — never the English string
 * itself. `PunchoutModal` renders it beside its own translated Status label,
 * so a hardcoded English map here was an untranslated badge one word away
 * from a translated one.
 *
 * Keyed on the union (not `string`, as the label map used to be) so a new
 * session status is a compile error rather than a badge that falls through to
 * its raw enum value; `catalog.test.ts` proves every key exists in the
 * catalogue. `PunchoutSession.status` is still a bare string on the wire —
 * `punchoutStatusLabelKey` is the tolerant accessor for it.
 */
export const PUNCHOUT_STATUS_LABEL_KEYS: Record<PunchoutSessionStatus, MessageKey> = {
	pending: 'catalogs.punchout.status.pending',
	returned: 'catalogs.punchout.status.returned',
	converted: 'catalogs.punchout.status.converted',
	expired: 'catalogs.punchout.status.expired',
	cancelled: 'catalogs.punchout.status.cancelled'
};

/** The message key for a session status, or `null` for an unrecognised one. */
export function punchoutStatusLabelKey(status: string): MessageKey | null {
	return PUNCHOUT_STATUS_LABEL_KEYS[status as PunchoutSessionStatus] ?? null;
}

// Badge tone per session status. Only `returned` and `converted` ever carried a
// colour of their own; everything else shared one grey tint, and still does.
export const PUNCHOUT_STATUS_TONES: Record<string, BadgeTone> = {
	pending: 'muted',
	returned: 'success',
	converted: 'accent',
	expired: 'muted',
	cancelled: 'muted'
};

/**
 * `PunchoutSession.status` is a string off the wire (the adapter may report a
 * provider status this union doesn't name), so the tone is read through a
 * tolerant accessor — mirroring how the label is read. An unrecognised status
 * gets the same grey the pill's base rule always gave it.
 */
export function punchoutStatusTone(status: string): BadgeTone {
	return PUNCHOUT_STATUS_TONES[status] ?? 'muted';
}

export interface PunchoutStartResponse {
	session_id: string;
	buyer_cookie: string;
	start_url: string;
	status: string;
	provider: string;
}

export interface PunchoutCartItem {
	description: string;
	sku: string | null;
	quantity: number | null;
	unit_price: MoneyAmount;
	uom: string | null;
	currency: string;
}

export interface PunchoutSession {
	id: string;
	catalog_id: string;
	buyer_cookie: string;
	status: string;
	requested_by_user_id: string;
	start_url: string | null;
	provider: string | null;
	cart_items: PunchoutCartItem[];
	cart_total: MoneyAmount;
	currency: string;
	returned_at: string | null;
	converted_requisition_id: string | null;
	created_at: string;
	updated_at: string;
}

export interface PunchoutConvertResponse {
	session_id: string;
	requisition_id: string;
	requisition_number: string;
	/** The created requisition's money total (`float` on the wire). */
	total: MoneyAmount;
	created: boolean;
}

File: backend/app/schemas/catalog.py
"""Pydantic request/response schemas for the catalogs + guided-buying router.

Money convention (mirrors ``schemas/expense.py`` / ``schemas/contract.py``):
request fields are typed ``Decimal | None`` for exactness on the way in;
response/list fields serialise money as ``float | None`` (the router does
``float(...)``). Never ``float`` on a column or in-memory total.

Catalogs are configuration-like (supplier / internal catalogs + their items);
``is_preferred`` steers guided buying. ``GuidedBuyingSuggestion`` is the
read-only steering surface — it ranks preferred sources, surfaces in-contract
vendors, and returns matching active catalog items so a buyer is pointed at the
right vendor / contract / catalog line before raising a requisition.
"""

from decimal import Decimal

from pydantic import BaseModel, Field

from app.api.pagination import PageMeta
from app.models.procurement import CatalogType

# ---------------------------------------------------------------------------
# Catalog items
# ---------------------------------------------------------------------------


class CatalogItemBase(BaseModel):
    sku: str | None = Field(default=None, max_length=100)
    name: str = Field(..., max_length=255)
    description: str | None = None
    # Digits match `catalog_items.unit_price` Numeric(15, 2). A catalog list
    # price is never negative.
    unit_price: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str = Field(default="USD", max_length=3)
    uom: str | None = Field(default=None, max_length=20)
    vendor_id: str | None = None
    gl_account_id: str | None = None
    category: str | None = Field(default=None, max_length=100)
    is_active: bool = True


class CatalogItemCreate(CatalogItemBase):
    pass


class CatalogItemUpdate(BaseModel):
    """PATCH — every field optional."""

    sku: str | None = Field(default=None, max_length=100)
    name: str | None = Field(default=None, max_length=255)
    description: str | None = None
    unit_price: Decimal | None = Field(default=None, ge=0, max_digits=15, decimal_places=2)
    currency: str | None = Field(default=None, max_length=3)
    uom: str | None = Field(default=None, max_length=20)
    vendor_id: str | None = None
    gl_account_id: str | None = None
    category: str | None = Field(default=None, max_length=100)
    is_active: bool | None = None


class CatalogItemResponse(BaseModel):
    id: str
    catalog_id: str
    sku: str | None
    name: str
    description: str | None
    unit_price: float | None
    currency: str
    uom: str | None
    vendor_id: str | None
    gl_account_id: str | None
    category: str | None
    is_active: bool
    created_at: str
    updated_at: str


# ---------------------------------------------------------------------------
# Catalogs
# ---------------------------------------------------------------------------


class CatalogBase(BaseModel):
    name: str = Field(..., max_length=255)
    catalog_type: CatalogType = CatalogType.internal
    vendor_id: str | None = None
    # Punch-out site URL — the supplier's hosted-catalog endpoint a buyer punches
    # out to. Live cXML/OCI round-trips are implemented (see
    # backend/docs/procurement-catalogs.md): a `punchout` catalog starts a
    # PunchoutSession against this URL.
    punchout_url: str | None = Field(default=None, max_length=500)
    is_active: bool = True
    is_preferred: bool = False
    description: str | None = None


class CatalogCreate(CatalogBase):
    pass


class CatalogUpdate(BaseModel):
    """PATCH — every field optional."""

    name: str | None = Field(default=None, max_length=255)
    catalog_type: CatalogType | None = None
    vendor_id: str | None = None
    punchout_url: str | None = Field(default=None, max_length=500)
    is_active: bool | None = None
    is_preferred: bool | None = None
    description: str | None = None


class CatalogResponse(BaseModel):
    id: str
    name: str
    catalog_type: str
    vendor_id: str | None
    punchout_url: str | None
    is_active: bool
    is_preferred: bool
    description: str | None
    item_count: int = 0
    items: list[CatalogItemResponse] = Field(default_factory=list)
    created_at: str
    updated_at: str


class CatalogListResponse(PageMeta):
    items: list[CatalogResponse]
    total: int


# ---------------------------------------------------------------------------
# Guided buying — read-only steering surface
# ---------------------------------------------------------------------------


class GuidedBuyingVendor(BaseModel):
    """A vendor surfaced as a preferred / in-contract source for a buyer."""

    vendor_id: str
    vendor_name: str
    # Why this vendor is recommended (ranked highest first):
    #   "preferred_catalog" — the vendor owns an active, preferred catalog
    #   "active_contract"   — the vendor has an active contract on file
    reasons: list[str] = Field(default_factory=list)
    # Linked active contract (if any) so the requisition can attach it.
    contract_id: str | None = None
    contract_number: str | None = None
    # The vendor's preferred catalog (if any) so the buyer can browse it.
    catalog_id: str | None = None
    catalog_name: str | None = None


class GuidedBuyingItem(BaseModel):
    """A matching active catalog item the buyer should prefer."""

    catalog_item_id: str
    catalog_id: str
    catalog_name: str
    sku: str | None
    name: str
    unit_price: float | None
    currency: str
    uom: str | None
    vendor_id: str | None
    category: str | None
    # True when the owning catalog is flagged preferred (ranked first).
    is_preferred: bool


class GuidedBuyingSuggestion(BaseModel):
    """Read-only guided-buying result. Steers a buyer to preferred vendors,
    in-contract vendors, and matching catalog lines for the given criteria
    (``category`` / ``vendor_id`` / free-text ``q``). Deterministic, no LLM."""

    preferred_vendors: list[GuidedBuyingVendor] = Field(default_factory=list)
    in_contract_vendors: list[GuidedBuyingVendor] = Field(default_factory=list)
    items: list[GuidedBuyingItem] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# Punch-out sessions — live cXML/OCI round-trip
# ---------------------------------------------------------------------------


class PunchoutStartResponse(BaseModel):
    """Result of ``POST /catalogs/{id}/punchout/start``.

    ``start_url`` is the supplier start page the buyer's browser visits;
    ``session_id`` + ``buyer_cookie`` correlate the eventual returned cart.
    """

    session_id: str
    buyer_cookie: str
    start_url: str
    status: str
    provider: str


class PunchoutCartItemResponse(BaseModel):
    """One returned cart line. Money serialises as ``float`` (out) per the file
    convention; the persisted value is exact ``Numeric``/``Decimal``."""

    description: str
    sku: str | None = None
    quantity: float | None = None
    unit_price: float | None = None
    uom: str | None = None
    currency: str = "USD"


class PunchoutSessionResponse(BaseModel):
    """A punch-out session view (start state and, once returned, the cart)."""

    id: str
    catalog_id: str
    buyer_cookie: str
    status: str
    requested_by_user_id: str
    start_url: str | None
    provider: str | None
    cart_items: list[PunchoutCartItemResponse] = Field(default_factory=list)
    cart_total: float | None
    currency: str
    returned_at: str | None
    converted_requisition_id: str | None
    created_at: str
    updated_at: str


class PunchoutConvertResponse(BaseModel):
    """Result of converting a returned session's cart into a requisition.

    ``created`` is ``False`` on the idempotent replay path (the session was
    already converted) so the caller can tell a fresh conversion from a no-op."""

    session_id: str
    requisition_id: str
    requisition_number: str
    total: float
    created: bool


__all__ = [
    "CatalogType",
    "CatalogItemBase",
    "CatalogItemCreate",
    "CatalogItemUpdate",
    "CatalogItemResponse",
    "CatalogBase",
    "CatalogCreate",
    "CatalogUpdate",
    "CatalogResponse",
    "CatalogListResponse",
    "GuidedBuyingVendor",
    "GuidedBuyingItem",
    "GuidedBuyingSuggestion",
    "PunchoutStartResponse",
    "PunchoutCartItemResponse",
    "PunchoutSessionResponse",
    "PunchoutConvertResponse",
]
